File: euler_projects/p58/solution.c
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>


int n = 1;
int u = 1;
int v = 1;
int w = 1;
int z = 1;
int dim = 1;
int total_primes_diag = 0;
int diag_size = 1;

void update(){
    int new_n = n + 1;


    dim = 2 * new_n - 1;
    v = u + 1 + dim - 2;
    w = v + dim - 1;
    z = w + dim - 1;
    u = z + dim - 1;

    n = new_n;


};


int * sieve(int b){

    static int * primes;
    primes = (int *)malloc(sizeof(int) * (b+1));

    for(int i=0; i<=b; i++){
        primes[i] = 1;
    };

    primes[0] = 0;
    primes[1] = 0;

    int p = 2;
    while(p * p <= b){
        if(primes[p] == 1){
            int k = p * p;
            while(k <= b){
                primes[k] = 0;
                k += p;
            };
        };
        p += 1;
    };

    return(primes);

};



int main(){

    double ratio = 1.0;

    int b = 1000 * 1000 * 1000;
    int * primes = sieve(b);
    printf("Found primes until %d \n", b);

    while(ratio >= 0.1){
        update();


        if(primes[u] == 1){
            total_primes_diag += 1;
        };
        if(primes[v] == 1){
            total_primes_diag += 1;
        };
        if(primes[w] == 1){
            total_primes_diag += 1;
        };
        if(primes[z] == 1){
            total_primes_diag += 1;
        };

        diag_size += 4;

        ratio = (double) total_primes_diag / diag_size;

        if(n % 100 == 0){
            printf("n %d - u %d - v %d - w %d - z %d - ratio=%d/%d = %f \n",
            n, u, v, w, z, total_primes_diag, diag_size, ratio);
        };

    };

    printf("n %d - u %d - v %d - w %d - z %d - ratio=%d/%d = %f \n",
    n, u, v, w, z, total_primes_diag, diag_size, ratio);


};

File: euler_projects/p58/solution.py
# solution.py
import math


class Spiral:

    def __init__(self):
        self.n = 1
        self.u = 1
        self.v = 1
        self.w = 1
        self.z = 1
        self.dim = 1

        self.primes = [False, True]

        self.ratio = 1.0
        self.total_primes_diag = 0
        self.diag_size = 1

        self.prev_u = 2

    def update(self):

        new_n = self.n + 1

        self.prev_u = self.u

        self.dim = 2 * new_n - 1
        self.v = self.u + 1 + self.dim - 2
        self.w = self.v + self.dim - 1
        self.z = self.w + self.dim - 1
        self.u = self.z + self.dim - 1

        self.n = new_n

    @staticmethod
    def sieve(b):

        primes = [True] * (b + 1)
        primes[0], primes[1] = False, False

        for p in range(2, int(math.sqrt(b)) + 1):
            if primes[p]:
                k = p ** 2
                while k <= b:
                    primes[k] = False
                    k += p
        return primes

    def find(self):

        while self.ratio >= 0.1:
            self.update()
            self.primes = self.sieve(self.u)

            print 'n {} :: u {}, v {}, w {}, z {}'.format(self.n, self.u, self.v, self.w, self.z)

            for elt in [self.v, self.w, self.z, self.u]:
                self.total_primes_diag += (1 if self.primes[elt] else 0)

            self.diag_size += 4

            self.ratio = float(self.total_primes_diag) / self.diag_size

            print 'n={} / ratio={}/{}={}%'.format(self.n, self.total_primes_diag, self.diag_size, round(self.ratio, 2))
